src/c++/perf_analyzer/test_load_manager.cc

```diff
@@ -0,0 +1,380 @@
+// Copyright 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "command_line_parser.h"
+#include "doctest.h"
+#include "load_manager.h"
+#include "test_load_manager_base.h"
+
+namespace cb = triton::perfanalyzer::clientbackend;
+
+namespace triton { namespace perfanalyzer {
+
+class TestLoadManager : public TestLoadManagerBase, public LoadManager {
+ public:
+  ~TestLoadManager() = default;
+  TestLoadManager(PerfAnalyzerParameters params, bool is_sequence_model = false)
+      : TestLoadManagerBase(params, is_sequence_model),
+        LoadManager(
+            params.async, params.streaming, params.batch_size,
+            params.max_threads, params.sequence_length,
+            params.shared_memory_type, params.output_shm_size,
+            params.start_sequence_id, params.sequence_id_range,
+            params.string_length, params.string_data, params.zero_input,
+            params.user_data, GetParser(), GetFactory())
```

Comment:
Additionally, is there a way to ditch this `GetParser()` and `GetFactory()` design here? It's difficult to understand.

Can each test just overwrite the `parser_` and/or `factory_` data members in the test body (if needed) before the relevant testing method gets called? Rather than trying to "intercept and inject a custom parser/factor" at constructor time, as is being done here?

I think seeing the injection happen in the test body will make it easier to understand what data members/logic is relevant to the test.

Reply:
In general, I think code duplication is slightly more acceptable for testing purposes. I think tests are better when someone can read mainly the test body and understand what's being mocked/changed and what production code is being called/verified. Having some testing logic that is really common I think is good to pull out, but I think in general, the more layers of complication we add with testing class hierarchy and variable passing via constructors and other roundabout ways makes the tests more error prone and harder to understand.

Reply:
> Additionally, is there a way to ditch this `GetParser()` and `GetFactory()` design here? It's difficult to understand.

There is not. These are expected at construction time, and there is live code in the real constructor that uses them.